### ISSUE TITLE
System.ObjectDisposedException: Cannot access a disposed object

### ISSUE THREAD

**Issue by JakobTischler** [rmcrackan/Libation#58]
Using v5.1.10: 

I've had a handful of cases where I get the "Skip importing this book?" popup. Sometimes it happens when I convert multiple books at the same time, sometimes when only converting a single one.

Log:
```log
2021-07-19 14:18:40.581 +02:00 [INF] (at LibationWinForms.BookLiberation.ProcessorAutomationController+<BackupSingleBookAsync>d__0.MoveNext()) Begin BackupSingleBookAsync {"productId":"B01MZ9QRBW"}
2021-07-19 14:18:40.590 +02:00 [INF] (at FileLiberator.IProcessableExt+<ProcessBookAsync_NoValidation>d__3.MoveNext()) Begin ProcessBookAsync_NoValidation {"Title":"A Closed and Common Orbit: Wayfarers Series, Book 2","AudibleProductId":"B01MZ9QRBW","Locale":"us","Account":"j[...]b.t[...]r@g[...]l.c[...]m"}
2021-07-19 14:18:40.593 +02:00 [INF] (at LibationWinForms.BookLiberation.LogMe+<>c.<.ctor>b__9_0(System.Object, System.String)) Automated backup: Decrypt Step, Begin: [B01MZ9QRBW] A Closed and Common Orbit: Wayfarers Series, Book 2
2021-07-19 14:18:40.607 +02:00 [INF] (at InternalUtilities.AudibleApiActions.GetApiAsync(System.String, System.String, AudibleApi.ILoginCallback)) GetApiAsync. {"Username":"j[...]b.t[...]r@g[...]l.c[...]m","LocaleName":"us"}
2021-07-19 14:18:41.139 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logBegin()) Begin step 'Download and Convert Aaxc To M4b'
2021-07-19 14:18:41.139 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logBegin()) Begin step 'Step 1: Get Aaxc Metadata'
2021-07-19 14:18:42.526 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logEnd(System.Boolean, System.TimeSpan, System.Exception)) End step 'Step 1: Get Aaxc Metadata'. Success. Completed in 00:00:01
2021-07-19 14:18:42.527 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logBegin()) Begin step 'Step 2: Download Decrypted Audiobook'
2021-07-19 14:18:54.054 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logEnd(System.Boolean, System.TimeSpan, System.Exception)) End step 'Step 2: Download Decrypted Audiobook'. Success. Completed in 00:00:11
2021-07-19 14:18:54.054 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logBegin()) Begin step 'Step 3: Create Cue'
2021-07-19 14:18:54.054 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logEnd(System.Boolean, System.TimeSpan, System.Exception)) End step 'Step 3: Create Cue'. Success. Completed in 00:00:00
2021-07-19 14:18:54.055 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logBegin()) Begin step 'Step 4: Create Nfo'
2021-07-19 14:18:54.055 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logEnd(System.Boolean, System.TimeSpan, System.Exception)) End step 'Step 4: Create Nfo'. Success. Completed in 00:00:00
2021-07-19 14:18:54.056 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logBegin()) Begin step 'Step 5: Cleanup'
2021-07-19 14:18:54.056 +02:00 [INF] (at Dinah.Core.IO.FileExt.SafeDelete(System.String)) File successfully deleted: C:\Users\XXX\AppData\Local\Temp\Libation\DownloadsInProgress\A Closed and Common Orbit Wayfarers Series, Book 2 [B01MZ9QRBW].json
2021-07-19 14:18:54.074 +02:00 [INF] (at Dinah.Core.IO.FileExt.SafeDelete(System.String)) File successfully deleted: C:\Users\XXX\AppData\Local\Temp\Libation\DownloadsInProgress\A Closed and Common Orbit Wayfarers Series, Book 2 [B01MZ9QRBW].aaxc
2021-07-19 14:18:54.075 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logEnd(System.Boolean, System.TimeSpan, System.Exception)) End step 'Step 5: Cleanup'. Success. Completed in 00:00:00
2021-07-19 14:18:54.075 +02:00 [INF] (at Dinah.Core.StepRunner.BaseStep.logEnd(System.Boolean, System.TimeSpan, System.Exception)) End step 'Download and Convert Aaxc To M4b'. Success. Completed in 00:00:12
2021-07-19 14:18:54.121 +02:00 [INF] (at LibationWinForms.Form1.setBookBackupCounts(System.Collections.Generic.IEnumerable`1[[DataLayer.Book, DataLayer, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]])) Book counts. {"fullyBackedUp":15,"downloadedOnly":0,"noProgress":12,"pending":12,"statusStripText":"BACKUPS: No progress: 12  Encrypted: 0  Fully backed up: 15","menuItemText":"12 remaining"}
2021-07-19 14:18:54.124 +02:00 [INF] (at LibationWinForms.Form1.setPdfBackupCounts(System.Collections.Generic.IEnumerable`1[[DataLayer.Book, DataLayer, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]])) PDF counts. {"downloaded":1,"notDownloaded":1,"statusStripText":"|  PDFs: NOT d/l'ed: 1  Downloaded: 1","menuItemText":"1 remaining"}
2021-07-19 14:18:54.126 +02:00 [INF] (at LibationWinForms.BookLiberation.LogMe+<>c.<.ctor>b__9_0(System.Object, System.String)) Automated backup: Decrypt Step, Completed: [B01MZ9QRBW] A Closed and Common Orbit: Wayfarers Series, Book 2

2021-07-19 14:18:54.127 +02:00 [ERR] (at LibationWinForms.BookLiberation.LogMe+<>c.<.ctor>b__9_2(System.Object, System.ValueTuple`2[[System.Exception, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]])) Automated backup: error
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'RichTextBox'.
   at System.Windows.Forms.Control.CreateHandle()
   at System.Windows.Forms.TextBoxBase.CreateHandle()
   at System.Windows.Forms.Control.get_Handle()
   at System.Windows.Forms.RichTextBox.get_TextLength()
   at Dinah.Core.Windows.Forms.RichTextBoxExt.AppendColourText(RichTextBox box, Color color, String text) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core.WindowsDesktop\Windows.Forms\RichTextBoxExt.cs:line 10
   at Dinah.Core.Windows.Forms.RichTextBoxTextWriter.writeLine(String value) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core.WindowsDesktop\Windows.Forms\RichTextBoxTextWriter.cs:line 25
   at Dinah.Core.Windows.Forms.RichTextBoxTextWriter.<>c__DisplayClass4_0.<WriteLine>b__0(RichTextBox c) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core.WindowsDesktop\Windows.Forms\RichTextBoxTextWriter.cs:line 14
   at Dinah.Core.Windows.Forms.ControlExt.InvokeIfRequired[T](T c, Action`1 action) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core.WindowsDesktop\Windows.Forms\ControlExt.cs:line 40
   at Dinah.Core.Windows.Forms.RichTextBoxTextWriter.WriteLine(String value) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core.WindowsDesktop\Windows.Forms\RichTextBoxTextWriter.cs:line 14
   at Dinah.Core.IO.MultiTextWriter.WriteLine(String value) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core\_IO\MultiTextWriter.cs:line 18
   at System.IO.TextWriter.SyncTextWriter.WriteLine(String value)
   at Dinah.Core.IO.MultiTextWriter.WriteLine(String value) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core\_IO\MultiTextWriter.cs:line 18
   at System.IO.TextWriter.SyncTextWriter.WriteLine(String value)
   at Dinah.Core.IO.MultiTextWriter.WriteLine(String value) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Dinah.Core\Dinah.Core\_IO\MultiTextWriter.cs:line 18
   at System.IO.TextWriter.SyncTextWriter.WriteLine(String value)
   at System.Console.WriteLine(String value)
   at AaxDecrypter.AaxcDownloadConverter.Run() in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\AaxDecrypter\AaxcDownloadConverter.cs:line 78
   at FileLiberator.DownloadDecryptBook.<aaxToM4bConverterDecryptAsync>b__38_2() in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\FileLiberator\DownloadDecryptBook.cs:line 98
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at FileLiberator.DownloadDecryptBook.aaxToM4bConverterDecryptAsync(String cacheDir, String destinationDir, LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\FileLiberator\DownloadDecryptBook.cs:line 98
   at FileLiberator.DownloadDecryptBook.ProcessAsync(LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\FileLiberator\DownloadDecryptBook.cs:line 41
   at FileLiberator.IProcessableExt.TryProcessAsync(IProcessable processable, LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\FileLiberator\IProcessableExt.cs:line 61
   at FileLiberator.BackupBook.ProcessAsync(LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\FileLiberator\BackupBook.cs:line 38
   at FileLiberator.IProcessableExt.ProcessBookAsync_NoValidation(IProcessable processable, LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\FileLiberator\IProcessableExt.cs:line 53
   at FileLiberator.IProcessableExt.ProcessSingleAsync(IProcessable processable, LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\FileLiberator\IProcessableExt.cs:line 40
   at LibationWinForms.BookLiberation.BackupRunner.ProcessOneAsync(Func`2 func, LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2019\Libation\LibationWinForms\BookLiberation\ProcessorAutomationController.cs:line 424
2021-07-19 14:18:54.136 +02:00 [ERR] (at LibationWinForms.BookLiberation.LogMe+<>c.<.ctor>b__9_1(System.Object, System.String)) ERROR. All books have not been processed. Most recent book: processing failed
```

Edit: sidenote: Could you maybe also add the book title / filename to the "Skip importing" popup? The user doesn't necessarily know which book is meant.

**Comment by wtanksleyjr** [comment]
As a side comment, when converting 1500 books, stopping on the first error
to ask my permission to continue is really counterproductive :) .

-Wm

**Comment by Mbucari** [comment]
Yeah, Libation doesn't really support simultaneous backups. It could if we rejiggered the logging, but that's a lot of work I'm not interested in doing. Unless or until that's done, we should make individual book backups a dialog so only one book can be backed up at a time.

**Comment by JakobTischler** [comment]
> Unless or until that's done, we should make individual book backups a dialog so only one book can be backed up at a time.

Yeah, that would be my preference as well. Don't allow the user to do something that could possibly mess things up.

**Comment by rmcrackan** [comment]
I don't necessarily recommend doing multiple ad hoc downloads at a time but I feel like it should be made possible. Issues 57 and 58 are both bugs that I need to solve by detangling which logging goes to which place. I'll take a look. To be frank though, there are a lot of higher priority items on plate right now. Unless it's a quick fix, this won't be addressed any time soon.

**Comment by rmcrackan** [comment]
Thanks for the log file. I added a bug fix which should solve this problem. It will be available with the next release. If you see this after the next release, please reply on this thread.

@JakobTischler 

> Edit: sidenote: Could you maybe also add the book title / filename to the "Skip importing" popup? The user doesn't necessarily know which book is meant.

Please open a new issue. It's hard to juggle multiple features in the same thread

@wtanksleyjr

> stopping on the first error to ask my permission to continue is really counterproductive

Rationale for the current design: if I don't ask permission, I could take action that the user doesn't want. If I add yet another flag setting, I doubt anyone will use it and that area will get confusing. I'm open to suggestions (preferable also in a new thread).

**Comment by rmcrackan** [comment]
Fix available in [v5.3.2](https://github.com/rmcrackan/Libation/releases/tag/v5.3.2)
